### ISSUE TITLE
Bad lexical cast

### ISSUE THREAD

**Issue by george-githinji** [yatisht/usher#126]
What does this error mean? 
```
terminate called after throwing an instance of boost::wrapexcept<boost::bad_lexical_cast>

.. bad lexical cast: source type value could not be interpreted as target
```

**Comment by yatisht** [comment]
Hi George,

Could you please share your command?

-Yatish

**Comment by george-githinji** [comment]
Thank you for the quick response. The command is
`MatUtils introduce -i new_assignments.pb -s samples.tsv -o introductions.tsv -M dates_table.tsv`

Where `dates.tsv `is a two column table with `sample_id` and `date`. 

I did not know what the input table  with dates should look like and how to label the two columns or the expected date format.

**Comment by jmcbroome** [comment]
Hi George- thanks for the report. The expected formatting for date mirrors that of the standard sample name format in our trees, e.g. YYYY-MM-DD, so that's correct.

I suspect from reading this that your issue is the ambiguous date you're inserting (the XXXX-XX-XX). If you want to ignore samples where you don't know what the date is, you can simply exclude these samples from your date tsv, and they will not be considered when establishing the active date range for a given cluster. Clusters which have no valid dates will get a default value indicating that no date could be parsed. 

If you could edit your date file to exclude these ambiguous samples, check to see if that resolves the issue, and post again in this thread, I'd appreciate it!

Report aside, we appreciate the feedback on the lack of clarity for using this option in terms of expected formatting and how to handle ambiguity. I'll be updating the wiki with additional information on this point soon.

**Comment by george-githinji** [comment]
> Hi George- thanks for the report. The expected formatting for date mirrors that of the standard sample name format in our trees, e.g. YYYY-MM-DD, so that's correct.
> 
> I suspect from reading this that your issue is the ambiguous date you're inserting (the XXXX-XX-XX). If you want to ignore samples where you don't know what the date is, you can simply exclude these samples from your date tsv, and they will not be considered when establishing the active date range for a given cluster. Clusters which have no valid dates will get a default value indicating that no date could be parsed.
> 
> If you could edit your date file to exclude these ambiguous samples, check to see if that resolves the issue, and post again in this thread, I'd appreciate it!
> 
> Report aside, we appreciate the feedback on the lack of clarity for using this option in terms of expected formatting and how to handle ambiguity. I'll be updating the wiki with additional information on this point soon.

Thank you for the quick and detailed reply. Very much appreciated. 
1. Just wondering if the sample column name should have a specific name e.g. sample_id, sequence_id or tip or it does not matter? 
2. In regard to fasta headers, when i provided sampleid_yyyy-mm-dd, I get a "not-a-date-time" in column 5 and 6 of the results file. Does this mean that the date was not properly read and what are the consequences of this. It it introduce any uncertainties to the results?

**Comment by jmcbroome** [comment]
Yes Yatish, I've been thinking along similar lines and was prepping an update this morning to do exactly that, as well as a couple other updates to the routine. That PR will be open soon. The wiki has had some additions as well this morning.

George, it shouldn't matter, as long as something is in the header in the first column position for the date metadata file- it doesn't explicitly check for "sample_id" like it checks for "date"- as long as that first column does in fact contain sample identifiers. Can you elaborate on what you're trying to do with fasta headers? If your sample name includes the date, it expects formatting ala standard name practice in the public databases, where the date is the last field "|" delimited in XXXX-XX-XX format. Where are you passing in strings like "sampleid_yyyy-mm-dd"?

**Comment by george-githinji** [comment]
Many thanks again for the clarification. Apologies for the confusion  I meant the samples.tsv file NOT the fasta headers. 

This applies to instances where the introduce command infers the date from the entry file (column1) of the tsv file. 
In your example files I noticed that the "|" character has some space around it such that " | " but it is not clear how many characters or what type of whitespace are expected. 
1. When I format the date as "samplename|yyyy-mm-dd" the command throws a segmentation fault. 
2. When I include a whitespace around e.g. samplename | yyyy-mm-dd , the command throws a segmentation error,
3. When I provide just the sample name and no date, the command fails with an error. 
4. But, when I provide the name as samplename_yyyy-mm-dd the command runs and input is written to file, however columns 5 and 6 contain a "not-a-date-time".

**Comment by jmcbroome** [comment]
Hi George- to further clarify, you're encountering the above errors when you do not use -M, but instead pass into -s a file that looks like:

sample1|2021-08-10    my_region
sample2|2021-07-30    my_region

Or simply, for the single region case

sample1|2021-08-10
sample2|2021-07-30 

Without a header? A header line is expected for the metadata input -M but not for -s.

**Comment by george-githinji** [comment]
Correct and Yes I am getting the errors when not using the -M argument but instead pass into -s a file 
sample|2021-08-10 my_region

**Comment by jmcbroome** [comment]
Such an input should work, as I just tested on my own machine- I'm not able to replicate your segfaults. Is it possible for you to provide the actual files and precise commands you're using so I can replicate the above issue? Either attached here or sent to me email jmcbroom@ucsc.edu. And just to be sure, is your installation the most recent version?

**Comment by jmcbroome** [comment]
Also, to address your previous question about whether the date introduces uncertainties into the results, the date is only used as context and a way to sort the output (outside of a new option we're introducing soon, -l). Even if your dates are not being parsed correctly for whatever reason, this will only affect the order in which output clusters are printed, and not the inference of clusters themselves.
